www_src/pages/section-2/grid.jsx

```diff
@@ -0,0 +1,382 @@
+          newLayout.forEach(function (row, index) {
+            row.unshift(null);
+          });
+
+          newPageX++;
+        },
+        right: function () {
+          newLayout.forEach(function (row, index) {
+            row.push(null);
+          });
+        },
+        top: function () {
+          newLayout.unshift(buildEmptyRow(newLayout[0].length));
+          newPageY++;
+        },
+        bottom: function () {
+          newLayout.push(buildEmptyRow(newLayout[0].length));
+        }
+      }
+
+      expand[direction]();
+    }
+
+    // Expand grid if necessary
+    if (edgeDetect(event.x, event.y).length) {
+      edgeDetect(event.x, event.y).forEach(function (direction, index, array) {
+        expandGrid(direction);
+      });
+    }
+
+    this.setState({
+      layout: newLayout
+    }, function () {
+      self.calculateCameraState(newPageX, newPageY, self.displayState.pagesWide, true);
+      self.animateCamera(false);
+    });
+  },
+  componentDidUpdate: function () {
+    var elGrid = this.getDOMNode();
+
+    // Force fixed width & height on grid
+    elGrid.style.width = this.gridWidth + 'px';
+    elGrid.style.height = this.gridHeight + 'px';
+  },
+  componentDidMount: function () {
+    this.gridWidth = undefined;
+    this.gridHeight = undefined;
+
+    this.slotWidth = undefined;
+    this.slotHeight = undefined;
+
+    this.tilesPerRow = undefined;
+    this.tilesPerCol = undefined;
+
+    this.displayState = {};
+  },
+  // Determine if a slot has a neighboring page in any cardinal direction
+  hasNeighbors: function (x, y) {
+    var layout = this.state.layout;
+
+    if (x > 0 && layout[y][x - 1]) {
+      return true;
+    }
+
+    if (x < layout[0].length - 1 && layout[y][x + 1]) {
+      return true;
+    }
+
+    if (y > 0 && layout[y - 1][x]) {
+      return true;
+    }
+
+    if (y < layout.length - 1 && layout[y + 1][x]) {
+      return true;
+    }
+
+    return false;
+  },
+  setContainerDimensions: function (width, height) {
+    this.setState({
+      containerWidth: width,
+      containerHeight: height
+    });
+  },
+  animateCamera: function (animateZoom) {
+    if (typeof animateZoom !== 'boolean') {
+      animateZoom = false;
+    }
+
+    var elGrid = this.getDOMNode();
+
+    var scaleTransform = 'scale(' + (this.displayState.zoom || 1) + ')';
+    var previousScaleTransform = 'scale(' + (this.displayState.previousZoom || 1) + ')';
+
+    var translateTransform = ' translate3d(' + this.displayState.cameraX + 'px, ' + this.displayState.cameraY + 'px, 0)';
+    var previousTranslateTransform = ' translate3d(' + this.displayState.previousCameraX + 'px, ' + this.displayState.previousCameraY + 'px, 0)';
+
+    var finalTransform = scaleTransform;
+
+    if (typeof this.displayState.cameraX === 'number' && typeof this.displayState.cameraY === 'number') {
+      finalTransform += (' '  + translateTransform)
+    }
+
+    if (animateZoom) {
+      elGrid.classList.add('animated');
```

Comment:
this feels like not taking advantage of React. If this element needs new classes, those should get set in `render` by using something like

```
var cn = require("classnames"); // npm hosted helper lib
render: function() {
  ...
  var elGridClasses = cl({
    elGrid: true,
    animated: animateZoom,
    ...
  });
  return <div classNames={elGridClasses} ...
    ...
}
`